src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +138,129 @@ private void logTargetPath() {
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
+    private void sortCamerasByLastTargetTimestamp() {
+        VisionObservationSorter.quickSortByDoubleValue(aprilTagCameras, (aprilTagCamera) -> ((AprilTagCamera) aprilTagCamera).getLatestResultTimestampSeconds());
+    }
+
+    private void addVisionObservations() {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
+            if (!aprilTagCamera.hasNewResult())
+                continue;
+
+            addVisionObservation(
+                    getCameraEstimatedPose(aprilTagCamera),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
+        }
+    }
+
+    private Pose2d getCameraEstimatedPose(AprilTagCamera aprilTagCamera) {
         final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
         if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
             return null;
+        return robotPose;
+    }
 
-        return new PoseEstimator6328.VisionObservation(
-                robotPose,
-                aprilTagCamera.getLatestResultTimestampSeconds(),
-                aprilTagCamera.calculateStandardDeviations()
-        );
+    private void addVisionObservation(Pose2d estimatedPose, PoseEstimatorConstants.StandardDeviations standardDeviations, double timestamp) {
+        if (isVisionObservationTooOld(timestamp))
+            return;
+
+        final Pose2d odometrySample = samplePose(timestamp);
+        if (odometrySample == null)
+            return;
+
+        final Transform2d odometryPoseToSamplePoseTransform = new Transform2d(odometryPose, odometrySample);
+        final Pose2d estimatedPoseAtObservationTime = estimatedPose.plus(odometryPoseToSamplePoseTransform);
+
+        final Pose2d estimatedOdometryPose = estimatedPoseAtObservationTime.plus(odometryPoseToSamplePoseTransform.inverse());
+        this.estimatedPose = estimatedOdometryPose.plus(calculatePoseStandardDeviations(estimatedPoseAtObservationTime, standardDeviations));
     }
 
-    private void putAprilTagsOnFieldWidget() {
-        for (Map.Entry<Integer, Pose3d> entry : FieldConstants.TAG_ID_TO_POSE.entrySet()) {
-            final Pose2d tagPose = entry.getValue().toPose2d();
-            field.getObject("Tag " + entry.getKey()).setPose(tagPose);
+    private boolean isVisionObservationTooOld(double timestamp) {
+        try {
+            final double oldestEstimatedRobotPoseTimestamp = previousOdometryPoses.getInternalBuffer().lastKey() - PoseEstimatorConstants.POSE_BUFFER_SIZE_SECONDS;
```

Comment:
`oldestEstimatedPoseTimestamp`

Reply:
It's in a method that uses vision, so it needs to be clear that it's using the robot's pose, not the camera's